[IPFS get] Do not cache exceptions

franckc opened the pull request:
The back-end processing queue had a few failures caused by IPFS timeouts. All the retries failed with logs that look like this:
```
2020-08-02T04:49:47.198016519Z [INFO]  dshop.utils.handleLog: Event OfferCreated for listing Id 1-001-233
2020-08-02T04:49:47.201167573Z [DEBUG] dshop.utils.events: Upsert event...
2020-08-02T04:49:47.207408834Z [DEBUG] dshop.utils.events: Event exists
2020-08-02T04:49:47.210113302Z [INFO]  dshop.utils.handleLog: OfferCreated - 592 by 0xDF73aF150b8E446a6D39FDdc2CFA7Bf067B88936
2020-08-02T04:49:47.210283391Z [INFO]  dshop.utils.handleLog: IPFS Hash: QmYNkDPPDtX1QbDBrjLSBJwr66GabdDh6DHG6vXwmeYNoG
2020-08-02T04:49:47.215113744Z [INFO]  dshop.utils.handleLog: Using IPFS gateway https://ipfs.ogn.app for fetching offer data
2020-08-02T04:49:47.21534627Z [INFO]  dshop.utils.handleLog: Fetching offer data with hash QmYNkDPPDtX1QbDBrjLSBJwr66GabdDh6DHG6vXwmeYNoG
2020-08-02T04:49:47.21873529Z [ERROR] dshop.queues.eventsProcessor: Log handler failed: IPFS gateway timeout
```

Notice that based on log timestamps, fetching the offer data from IPFS returned with a timeout immediately, despite the timeout being set to 60sec... I checked the content for that hash is available by logging into a server container and fetching the content.

It turns out there is a subtlety in the way memoize works for async functions. It caches the Promise. Therefore, if the Promise gets rejected (due to an IPFS timeout), all subsequent calls to the function for the same IPFS hash would return the same exception as opposed to trying to fetch the content again. Which explains why the queue retries all failed.

Fixed by adding a wrapper method that clears the memoize cache in case of an exception.

Also cleaned up Jsdoc for this file.

franckc commented:
> Looks good. Did you just test it manually, or wondering if it's worth adding a simple test in here...

Yes I did test manually but let me add a quick unit test. Good to have!

mikeshultz commented on backend/utils/_ipfs.js:
```diff
@@ -186,20 +165,44 @@ async function getTextFn(gateway, hashAsBytes, timeoutMS) {
+ */
+async function getText(gateway, hashAsBytes, timeoutMS) {
+  try {
+    return getTextCache(gateway, hashAsBytes, timeoutMS)
```
I vaguely recall dealing with this before in graphql(and [replacing memoize](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/utils/memorize.js)).  Are you sure this will throw as you expect?

franckc replied:
Thank you for pointing this out. You were right, it did not throw an exception. Fixed and added unit test.

franckc commented:
- Reworked the memoize wrapper and moved it under @origin/utils/memoize
 - Refactored the code a bit so that we now have a single place @origin/ipfs/index where this code leaves as opposed to duplicating it under backend/src/util/_ipfs and @origin/ipfs

nick commented on packages/utils/package.json:
```diff
@@ -16,7 +16,8 @@
     "lodash": "^4.17.19",
-    "logplease": "^1.2.15"
+    "logplease": "^1.2.15",
+    "@origin/utils": "^0.1.0"
```
Depends on itself?

franckc replied:
Oops. Good catch. That was supposed to be in packages/ipfs/package.json. Fixed.
